File: 200-NumberOfIslands/200-NumberOfIslands.py
# Last updated: 10/8/2025, 9:51:59 PM
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        def isValid(row,col):
            return 0<=row<m and 0<=col<n and grid[row][col]=="1"



        def dfs(row,col):
            for dx,dy in directions:
                nextRow, nextCol=row+dy,col+dx
                if isValid(nextRow,nextCol) and (nextRow,nextCol) not in seen: 
                    seen.add((nextRow,nextCol))
                    dfs(nextRow,nextCol)




        directions=[(0,1),(1,0),(0,-1),(-1,0)]
        m=len(grid)
        n=len(grid[0])
        seen=set()
        ans=0

        for row in range(m):
            for col in range(n):
                if grid[row][col]=="1" and (row,col) not in seen:
                    ans+=1
                    seen.add((row,col))
                    dfs(row,col)
        return ans


        

File: 200-NumberOfIslands/200-NumberOfIslands.java
// Last updated: 10/8/2025, 9:52:01 PM
class Solution {
    private int m;
    private int n;
    public int[][]  dirs=new int[][]{{0,1},{0,-1},{-1,0},{1,0}};
            public int numIslands(char[][] grid) {
            int count=0;
            //numner rows and column..iterate in forloop..if we find ==1..cout ++;bfs()
            m= grid.length;
            n=grid[0].length;
            for(int i=0;i<m;i++){
                for(int j=0;j<n;j++){
                    if(grid[i][j]=='1'){
                        count++;
                        bfsGraph3(grid,i,j);
                    }
                }

            }

            return count;

        }

        public void bfsGraph3(char[][] grid,int r,int c){
            Queue<int[]> que=new LinkedList<>();
            
            que.add(new int[]{r,c});
            while ((!que.isEmpty())){
                int[] curr=que.poll();
                for(int[] dir: dirs){
                int nr=curr[0]+dir[0];
                int nc=curr[1]+dir[1];
                if(0<=nr && nr<m && 0<=nc && nc<n && grid[nr][nc]=='1'){
                grid[nr][nc]='0';
                que.offer(new int[]{nr,nc});
                }
                }
            }
        }
}